### PR TITLE
Keep dynamic batches about the same shape

### PR DESCRIPTION
**Pull request by stephenroller**
**Patch description**
This patch aims to improve distributed training with dynamic batches. In distributed training, we are always bottlenecked by our slowest worker, as we need to `sync()` on every parley and every SGD step.

*On main branch*:
Dynamic batching introduces variability into the training procedure. When building a dynamic batch, we start with a random example, and then fill the batch with similar length examples until we run out of space. One issue here is that the workers may choose differently sized batches. For example, worker A might build a batch with 100 short examples, while worker B might build a batch with only a 5 very long examples. Both workers will forward/backward on these batches and both use roughly the same amount of memory, so there is actually _less_ variability in their performance. Sounds great, right?

However, tokenization is one of the most expensive parts of training (and one major reason we pay a performance cost relative to fairseq). Once both workers finish the current batch they are working on, it will be necessary for them to tokenize the _next_ sentences of the conversations. Here lies the straggler issue: worker A will now need to tokenize 100 sentences, but worker B only needs to tokenize 5. We can reduce this variability by ensuring _all_ workers attempt to build similarly-shaped batches (making the same many-short or few-long decision at the same time), so they all have roughly the same number of items to tokenize.

*With this patch*:
This patch adds this heuristic to training. While we can't guarantee that all workers make the same choice of batch size, we can ensure that they begin building from the same position when starting with the sorted list:

https://github.com/facebookresearch/ParlAI/blob/6728a9ad3d3a36e7d8f9315d027f51500c14cf47/parlai/core/worlds.py#L1197-L1207

Note that we need a separate RNG for this, because not all workers will pick the exact same sized batch, just similarly sized ones. This means the rng usage above could cause them to get out of sync.

However there is an issue: this introduces a significant amount of determinism into the conversation selection process, and separate workers may then begin to select the exact same conversation over. This reduces our stochasticity, and hinders learning. To remedy this, we expand upon our recent changes (#2433, #2445), which stratifies evaluation across many workers by having each of the N workers take only every Nth episode. We expand this to also include training, so the data loaded into memory by each worker is separate, and therefore they can never pick different examples.

**Testing steps**
Correctness tests:
- Added a print statement to all workers (using 8 workers) to print their examples on each step. Before the stratification change, I confirmed that many workers were picking the same conversation. I introduced the stratification change in order to fix this, and verified this was the case.
- Added a print statement to all workers to print their batch sizes. Verified that workers were creating roughly equal batch lengths.
- Trained small models and verified the PPL looked right. Before the stratification change, there was a performance hit (though it was somewhat moderate)

Timing tests:
- Trained on `large_secret_internal_teacher` (a ChunkTeacher) for 300 seconds. Measured the number of updates.
- Trained on `-t convai2 -dt train` for 300 seconds. Measured number of updates
- Trained on `-t convai2 -dt train:valid` for 300 seconds. Measured the number of updates.

**Logs**
Command run (via srun)
```
python -u -m parlai.scripts.distributed_train  -m transformer/generator -esz 128 --ffn-size 512 --n-heads 4 --dict-tokenizer bytelevelbpe --dict-file zoo:blender/blender_3B/model.dict -tr 512 -bs 32 --skip-generation true -opt adam -lr 1e-4 -vmt ppl --warmup-updates 1000 -bs 32 -dynb full -nl 8 -ttim 300 --distributed-world-size GPUS  -dt DATATYPE -t TEACHER
```

Legend: upd (m) = updates on main branch; upd (b) = updates on this branch.

 GPUs | Teacher |     Datatype |     Upd (m) | Upd (b) 
----- | ------- | ------------ | ----------- | --------
8     |  secret | train:stream |       601   | **679** 
16    |  secret | train:stream |       581   | **667** 
32    |  secret | train:stream |       543   | **617** 
64    |  secret | train:stream |       500   | **584** 
8     | ConvAI2 | train:stream |       799   | **817** 
16    | ConvAI2 | train:stream |     **762** |   748   
8     | ConvAI2 | train        |       801   | **867** 
16    | ConvAI2 | train        |       738   | **829** 
       
The one where we regressed slightly is explained to I/O thrashing: because ConvAI2 is so small, using 16 GPUs caused it to be loaded about 12 times in the experiment. Solution is to not use streaming with small data.

### CODE REVIEW

File: parlai/core/teachers.py [hunk 1]
```diff
@@ -396,29 +405,6 @@ def next_example(self):
 
         return ex, epoch_done
 
-    def next_batch(self):
```
Comment (stephenroller) [inline parlai/core/teachers.py:399]: this was never called anywhere. Just deleting some old code I found.

**Comment by stephenroller** [comment]
Putting this back as a draft. It's turned out to be deviously trickly to get right. I think the Teacher rewrite may need to come first.

File: parlai/core/teachers.py [hunk 2]
```diff
@@ -356,10 +359,16 @@ def next_episode_idx(self, num_eps=None, loop=None):
         if loop is None:
             loop = self.training
         if self.random:
-            new_idx = random.randrange(num_eps)
+            new_idx = (
+                random.randrange(
+                    num_eps // self.dws + int(num_eps % self.dws > self.rank)
```
Comment (klshuster) [inline parlai/core/teachers.py:364]: what is this computation exactly?

Comment (stephenroller) [reply parlai/core/teachers.py:364]: we split across dws (distributed world size) workers.

if there are 1000 episodes and 16 workers, then 8 workers get 63 examples, and 8 workers get 62.

8 * 62 + 8 * 63 = 1000.

The > rank mess is identifying who gets the the 63 (the first workers) and who gets the 62 (the last workers)

File: parlai/core/teachers.py [hunk 3]
```diff
@@ -705,6 +692,9 @@ def share(self):
             'cands': self.cands,
             'image_loader': self.image_loader,
         }
+        if hasattr(self, '_num_examples_cache'):
+            shared['num_examples_cache'] = (self._num_examples_cache,)
```
[the remark this answers is not in the document]
Comment (stephenroller) [reply parlai/core/teachers.py:696]: bug

**Comment by klshuster** [comment]
@stephenroller is this something we should look back into? from your experiments it seemed to have worked pretty well in terms of speedup

**Comment by stephenroller** [comment]
Yeah I'd very much like to land this one day, but it's just not any priority. It requires reworking teachers.

**Comment by klshuster** [comment]
applying `donotreap`

**Comment by stephenroller** [comment]
Finding a different way to do this.

**Pull request closed without merging**